lib/spack/spack/test/spec_yaml.py

```diff
@@ -204,6 +211,72 @@ def test_ordered_read_not_required_for_consistent_dag_hash(
         assert spec.full_hash() == round_trip_reversed_json_spec.full_hash()
 
 
+@pytest.mark.parametrize("module", [
+    spack.spec,
+    spack.architecture,
+    spack.version,
+])
+def test_hashes_use_no_python_dicts(module):
+    """Coarse check to make sure we don't use dicts in Spec.to_node_dict().
+
+    Pyton dicts are not guaranteed to iterate in a deterministic order
+    (at least not in all python versions) so we need to uses lists and
+    syaml_dicts.  syaml_dicts are ordered and ensure that hashes in Spack
+    are deterministic.
+
+    This test is intended to handle cases that are not covered by the
+    consistency checks above, or that would be missed by a dynamic check.
+    This test traverses the ASTs of functions that are used in our hash
+    algorithms, finds instances of dictionaries being constructed, and
+    prints out the line numbers where they occur.
+
+    """
+    class FindFunctions(ast.NodeVisitor):
+        """Find a function definition called to_node_dict."""
+        def __init__(self):
+            self.nodes = []
+
+        def visit_FunctionDef(self, node):  # noqa
+            if node.name in ("to_node_dict", "to_dict", "to_dict_or_value"):
+                self.nodes.append(node)
```

Comment:
We have several other `to_dict` methods defined in modules other than the ones tested. Skimming through them they don't appear to enter the hashing algorithm. Is it worth considering renaming the ones that are used to compute the hash `to_ordered_dict` etc. (that should make clear to anybody modifying the methods what is the expected return value semantic)?

Reply:
I like @scheibelp's suggestion to just later refactor things so that we sort the dict values *in* the hashing algorithm - we'll do that later but not in this PR